gateway/gateway.go

```diff
@@ -162,50 +168,115 @@ func (g *gateway) Reconfigure(ctx context.Context, deps resource.Dependencies, c
 
 	errCode := C.setUpGateway(C.int(cfg.Bus))
 	if errCode != 0 {
-		return fmt.Errorf("failed to start the gateway %d", errCode)
+		strErr := parseErrorCode(int(errCode))
+		return fmt.Errorf("failed to set up the gateway: %s", strErr)
 	}
 
 	g.started = true
-	g.receivePackets()
+	g.workers.Add(g.receivePackets)
 
 	return nil
 }
 
-func (g *gateway) receivePackets() {
+func parseErrorCode(errCode int) string {
+	switch errCode {
+	case 1:
+		return "error setting the board config"
+	case 2:
+		return "error setting the radio frequency config for radio 0"
+	case 3:
+		return "error setting the radio frequency config for radio 1"
+	case 4:
+		return "error setting the intermediate frequency chain config"
+	case 5:
+		return "error starting the gateway"
+	default:
+		return "unknown error"
+	}
+}
+
+// startCLogging starts the goroutine to capture C logs into the logger.
+// If loggingRoutineStarted indicates routine has already started, it does nothing.
+func (g *gateway) startCLogging(ctx context.Context) {
+	loggingState, ok := loggingRoutineStarted[g.Name().Name]
+	if !ok || !loggingState {
+		g.logger.Debug("Starting c logger background routine")
+		g.workers = utils.NewBackgroundStoppableWorkers(g.captureCOutputToLogs)
+		loggingRoutineStarted[g.Name().Name] = true
+	}
+}
+
+// captureOutput is a background routine to capture C's stdout and log to the module's logger.
+// This is necessary because the sx1302 library only uses printf to report errors.
+func (g *gateway) captureCOutputToLogs(ctx context.Context) {
+	// Need to disable buffering on stdout so C logs can be displayed in real time.
+	C.disableBuffering()
+
+	stdoutR, stdoutW, err := os.Pipe()
+	if err != nil {
+		g.logger.Errorf("unable to create pipe for C logs")
+		return
+	}
+	// Redirect C's stdout to the write end of the pipe
+	C.redirectToPipe(C.int(stdoutW.Fd()))
+	scanner := bufio.NewScanner(stdoutR)
+
+	//nolint:errcheck
+	defer stdoutR.Close()
+	//nolint:errcheck
+	defer stdoutW.Close()
+
+	// loop to read lines from the scanner and log them
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			if scanner.Scan() {
+				line := scanner.Text()
+				if strings.Contains(line, "ERROR") {
+					g.logger.Error(line)
+				} else {
+					g.logger.Debug(line)
+				}
```

Comment:
[nit, optional] A case structure is always easier to read to me than an if-else statement.

Reply:
sure can change!